code/HabitTracker/app/src/main/java/com/example/habittracker/DatabaseManager.java

```diff
@@ -240,4 +251,162 @@ public void onFailure(@NonNull Exception e) {
+                if (task.isSuccessful()) {
+                    ArrayList<Habit> habitArray = new ArrayList<>();
+                    for (QueryDocumentSnapshot doc : task.getResult()) {
+                        Log.d("",""+doc.getData().get("whatDays")+"-----"+doc.getData().get("dateStarted"));
+                        ArrayList<String> daysArray = (ArrayList<String>) doc.getData().get("whatDays");
+                        ArrayList<Long> dateArray = (ArrayList<Long>) doc.getData().get("dateStarted");
+                        Calendar cal = Calendar.getInstance();
+                        if(dateArray == null || daysArray == null){
+                            continue;
+                        }
+                        cal.set(dateArray.get(0).intValue(),dateArray.get(1).intValue()-1,dateArray.get(2).intValue());
+                        Date date = cal.getTime();
+                        habitArray.add(new Habit(
+                                doc.getId(),
+                                (String)doc.getData().get("reason"),
+                                date,
+                                daysArray.toArray(new String[daysArray.size()])
+                        ));
+                    }
+                    callback.onCallbackSuccess(habitArray);
+                }
+                else{
+                    callback.onCallbackFailed();
+                }
+            }
+        });
+    }
+
+    /**
+     * This method return an array list of all habit events for a habit using a callback function.
+     * @param user
+     * @param habit
+     * @param callback
+     */
+    public void getAllHabitEvents(String user, Habit habit, HabitEventListCallback callback) {
+        // Users -> userid (key) -> Habits -> habitTitle (key) -> HabitEvents
+        CollectionReference doc = usersColRef
+                .document(user)
+                .collection(habitsColName)
+                .document(habit.getTitle())
+                .collection(habitEventsColName);
+        doc.get().addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<QuerySnapshot> task) {
+                if (task.isSuccessful()) {
+                    ArrayList<HabitEvent> eventArray = new ArrayList<>();
+                    for (QueryDocumentSnapshot doc : task.getResult()) {
+                        Date date;
+                        try {
+                             date = new SimpleDateFormat("yyyy-MM-dd").parse((String)doc.getData().get("startDate"));
+                        } catch (Exception e) {
+                            date = new Date();
+                        }
+                        eventArray.add(new HabitEvent(
+                                habit,
+                                (String)doc.getData().get("comment"),
+                                date,
+                                (String)doc.getData().get("comment"),
```

Comment:
This is good for now. I'm sure you are aware that we need to get the location field in the future.

Reply:
yes alot of the major classes are gonna change when everyone else PR in so things like that are gonna change